usdValidation: usdShade: NormalMapTextureValidator

beersandrew opened the pull request:
### Description of Change(s)
Added an implementation of NormalMapTexture validation rules

### Link to proposal ([if applicable](https://openusd.org/release/contributing_to_usd.html#step-1-get-consensus-for-major-changes))


### Fixes Issue(s)
- https://github.com/orgs/PixarAnimationStudios/projects/3?pane=issue&itemId=65425006

Note: this PR is a replacement for https://github.com/PixarAnimationStudios/OpenUSD/pull/3359

### Checklist

[X] I have created this PR based on the dev branch

[X] I have followed the [coding conventions](https://openusd.org/release/api/_page__coding__guidelines.html)

[X] I have added unit tests that exercise this functionality (Reference: 
[testing guidelines](https://openusd.org/release/api/_page__testing__guidelines.html))

[X] I have verified that all unit tests pass with the proposed changes

[X] I have submitted a signed Contributor License Agreement (Reference: 
[Contributor License Agreement instructions](https://openusd.org/release/contributing_to_usd.html#contributor-license-agreement))

beersandrew committed:
feat: added an implementation for the NormalMapTexture validator

- added all necessary files to move  the NormalMapTexture validator from complianceChecker.py to the new Validation Framework

beersandrew committed:
fix: making sure lines are less than 80 characters wide

jesschimein commented:
Filed as internal issue #USD-10485

jesschimein commented:
/AzurePipelines run

tallytalwar commented on pxr/usdValidation/usdShadeValidators/plugInfo.json:
```diff
@@ -17,7 +17,10 @@
                         "doc": "All properties named 'material:binding' or in that namespace should be relationships."
-                    }, 
+                    },
+                    "NormalMapTextureValidator" : {
```
I think this should have a schemaTypes of UsdShadeShader associated with it.

tallytalwar commented on pxr/usdValidation/usdShadeValidators/validators.cpp:
```diff
@@ -538,6 +540,245 @@ _SubsetsMaterialBindFamily(const UsdPrim &usdPrim)
+    const UsdShadeShader shader(usdPrim);
+
+    TfToken shaderId;
+    TfToken UsdPreviewSurface("UsdPreviewSurface");
```
const

tallytalwar commented on pxr/usdValidation/usdShadeValidators/validators.cpp:
```diff
@@ -538,6 +540,245 @@ _SubsetsMaterialBindFamily(const UsdPrim &usdPrim)
+    }
+
+    TfToken sourceShaderId;
+    TfToken UsdUVTexture("UsdUVTexture");
```
const

tallytalwar commented on pxr/usdValidation/usdShadeValidators/validators.cpp:
```diff
@@ -538,6 +540,245 @@ _SubsetsMaterialBindFamily(const UsdPrim &usdPrim)
+        return {};
+    }
+
+    const auto getInputValue = [](const UsdShadeShader &inputShader,
```
Coding conventions to name private functions/lamdas: _GetInputValue

tallytalwar commented on pxr/usdValidation/usdShadeValidators/validators.cpp:
```diff
@@ -538,6 +540,245 @@ _SubsetsMaterialBindFamily(const UsdPrim &usdPrim)
+    }
+
+    const auto getInputValue = [](const UsdShadeShader &inputShader,
+        const TfToken &token, auto& outputValue) -> bool {
```
Our coding conventions use a pointer for an outValue.

tallytalwar commented on pxr/usdValidation/usdShadeValidators/validators.cpp:
```diff
@@ -538,6 +540,245 @@ _SubsetsMaterialBindFamily(const UsdPrim &usdPrim)
+                           sourcePrim.GetPath().GetText(), assetPath.c_str()));
+    }
+
+    auto textureIs8Bit = [](std::string resolvedPath) {
```
_TextureIs8Bit, same reason as above.

tallytalwar commented on pxr/usdValidation/usdShadeValidators/validators.cpp:
```diff
@@ -538,6 +540,245 @@ _SubsetsMaterialBindFamily(const UsdPrim &usdPrim)
+    }
+
+    TfToken colorSpace;
+    TfToken rawColorSpace("raw");
```
const

tallytalwar commented on pxr/usdValidation/usdShadeValidators/validators.cpp:
```diff
@@ -538,6 +540,245 @@ _SubsetsMaterialBindFamily(const UsdPrim &usdPrim)
+    };
+
+    SdfAssetPath textureAssetPath;
+    bool valueForFileExists = getInputValue(sourceShader, TfToken("file"),
```
nit: Might make sense to do a const TfToken FileToken("file") and use that.

Also FYI, If you find that some tokens are being used multiple times across validators, it might also make sense to use file static tokens: https://openusd.org/dev/api/static_tokens_8h.html#details. Example: https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/imaging/hd/coordSys.cpp#L14-L18

tallytalwar commented on pxr/usdValidation/usdShadeValidators/validators.cpp:
```diff
@@ -538,6 +540,245 @@ _SubsetsMaterialBindFamily(const UsdPrim &usdPrim)
+    TfToken colorSpace;
+    TfToken rawColorSpace("raw");
+    bool valueForColorSpaceExists =
+        getInputValue(sourceShader, TfToken("sourceColorSpace"), colorSpace);
```
Same note as above:
const TfToken sourceColorSpaceToken("sourceColorSpace")

tallytalwar commented on pxr/usdValidation/usdShadeValidators/validators.cpp:
```diff
@@ -538,6 +540,245 @@ _SubsetsMaterialBindFamily(const UsdPrim &usdPrim)
+    }
+
+    GfVec4f bias;
+    bool valueForBiasExists = getInputValue(sourceShader, TfToken("bias"),
```
const TfToken bias("bias")

tallytalwar commented on pxr/usdValidation/usdShadeValidators/validators.cpp:
```diff
@@ -538,6 +540,245 @@ _SubsetsMaterialBindFamily(const UsdPrim &usdPrim)
+        bias);
+
+    GfVec4f scale;
+    bool valueForScaleExists = getInputValue(sourceShader, TfToken("scale"),
```
const TfToken scale("scale")

tallytalwar commented on pxr/usdValidation/usdShadeValidators/testenv/testUsdShadeValidators.cpp:
```diff
@@ -59,8 +61,10 @@ TestUsdShadeValidators()
+    TF_AXIOM(metadata.size() == 8);
     for (const UsdValidationValidatorMetadata &metadata : metadata) {
+
+        std::string name = metadata.name.GetString();
```
Why convert the name token to string here and then put it in the token set again?

tallytalwar replied:
Not using this name now I guess?

tallytalwar commented on pxr/usdValidation/usdShadeValidators/testenv/testUsdShadeValidators.cpp:
```diff
@@ -554,13 +558,223 @@ TestUsdShadeEncapsulationRulesValidator()
+    normalInput.ConnectToSource(
+        SdfPath("/RootMaterial/NormalTexture.outputs:rgb"));
+
+    // Verify invalid bias & scale error, they should exists and do
+    // not exist at this point.
```
What do you mean by "they should exists and do not exist at this point"?

tallytalwar replied:
Still have this around? Could you please clarify or update the doc?

tallytalwar commented on pxr/usdValidation/usdShadeValidators/testenv/testUsdShadeValidators.cpp:
```diff
@@ -554,13 +558,223 @@ TestUsdShadeEncapsulationRulesValidator()
+    "usdShadeValidators:NormalMapTextureValidator.NonCompliantBiasAndScale");
+    std::string expectedErrorMsg =
+        TfStringPrintf("UsdUVTexture prim <%s> reads 8 bit Normal Map "
+                           "@./normalMap.jpg@, which requires that "
+                           "inputs:scale be set to (2, 2, 2, 1) and "
+                           "inputs:bias be set to (-1, -1, -1, 0) for proper "
+                           "interpretation as per the UsdPreviewSurface and "
+                           "UsdUVTexture docs.",
```
there is extra indentation here.

```cpp
TfStringPrintf("UsdUVTexture prim <%s> reads 8 bit Normal Map "
               "@./normalMap.jpg@, which requires that "
               "..."
               "...",
```

tallytalwar commented on pxr/usdValidation/usdShadeValidators/testenv/testUsdShadeValidators.cpp:
```diff
@@ -554,13 +558,223 @@ TestUsdShadeEncapsulationRulesValidator()
+
+    // Verify invalid bias & scale error, they should exists and do
+    // not exist at this point.
+    UsdValidationErrorVector errors = validator->Validate(
+        usdPreviewSurfaceShaderPrim);
+    TfToken expectedErrorIdentifier(
+    "usdShadeValidators:NormalMapTextureValidator.NonCompliantBiasAndScale");
+    std::string expectedErrorMsg =
+        TfStringPrintf("UsdUVTexture prim <%s> reads 8 bit Normal Map "
+                           "@./normalMap.jpg@, which requires that "
+                           "inputs:scale be set to (2, 2, 2, 1) and "
+                           "inputs:bias be set to (-1, -1, -1, 0) for proper "
+                           "interpretation as per the UsdPreviewSurface and "
+                           "UsdUVTexture docs.",
+                           usdUvTextureShader.GetPath().GetText());
+    ValidateError(errors,
+        expectedErrorIdentifier,
+        usdUvTextureShader.GetPath(),
+        expectedErrorMsg);
```
You can scope this such that the expected* variables are const, as is done in other testing methods here.
```cpp
{
    const UsdValidationErrorVector errors = validator->...;
    const TfToken expectedErrorIdentifier("...");
    const std::string expectedErrorMsg = TfStringPrintf("...");
    ValidateErrors(errors, expectedErrorIdentifier, usdUvTextureShader.GetPath(), expectedErrorMsg);
}
```

tallytalwar replied:
Same for below.

tallytalwar commented on pxr/usdValidation/usdShadeValidators/testenv/testUsdShadeValidators.cpp:
```diff
@@ -554,13 +558,223 @@ TestUsdShadeEncapsulationRulesValidator()
+    TF_AXIOM(errors[0].GetSites()[0].GetPrim().GetPath() ==
+             expectedPrimPath);
+    TF_AXIOM(errors[0].GetMessage() == expectedErrorMsg);
+}
```
Thanks for this; Might make sense to also update other testing methods here to use this, could you do that also, as you are introducing this common method?

beersandrew replied:
Happy to do this. Does it make sense to bring this helper function up to `usdValidation` and then reuse it in all domain specific validators?

If so, do you want that on this PR or on a separate PR?

tallytalwar replied:
Hmm good thought, I feel based on other unit tests in all of openUsd, a pattern established is to have helper like these stick with each unit test. So I think its fine to have this helper function for each test file.

tallytalwar commented on pxr/usdValidation/usdShadeValidators/testenv/testUsdShadeValidators.cpp:
```diff
@@ -554,13 +558,223 @@ TestUsdShadeEncapsulationRulesValidator()
+    "usdShadeValidators:NormalMapTextureValidator.NonCompliantBiasValues");
+    expectedErrorMsg =
+        TfStringPrintf("UsdUVTexture prim <%s> reads an 8 bit Normal "
+                            "Map, but has non-standard inputs:bias value of "
+                            "(%.6g, %.6g, %.6g, %.6g). inputs:bias must be "
+                            "set to [-1,-1,-1,0] so as to fulfill the "
+                            "requirements of the normals to be in tangent "
+                            "space of [(-1,-1,-1), (1,1,1)] as documented in "
+                            "the UsdPreviewSurface and UsdUVTexture docs.",
+                            usdUvTextureShader.GetPath().GetText(),
+                            nonCompliantVector[0], nonCompliantVector[1],
+                            nonCompliantVector[2], nonCompliantVector[3]);
```
indentation

tallytalwar replied:
Same below.

tallytalwar commented on pxr/usdValidation/usdShadeValidators/testenv/testUsdShadeValidators.cpp:
```diff
@@ -554,13 +558,223 @@ TestUsdShadeEncapsulationRulesValidator()
+    const GfVec4f compliantScale = GfVec4f(2, 2, 2, 1);
+    scaleInput.Set(compliantScale);
+
+    // Verify the non-compliant bias value error occurs.
+    errors = validator->Validate(usdPreviewSurfaceShaderPrim);
+    expectedErrorIdentifier = TfToken(
+    "usdShadeValidators:NormalMapTextureValidator.NonCompliantBiasValues");
+    expectedErrorMsg =
+        TfStringPrintf("UsdUVTexture prim <%s> reads an 8 bit Normal "
+                            "Map, but has non-standard inputs:bias value of "
+                            "(%.6g, %.6g, %.6g, %.6g). inputs:bias must be "
+                            "set to [-1,-1,-1,0] so as to fulfill the "
+                            "requirements of the normals to be in tangent "
+                            "space of [(-1,-1,-1), (1,1,1)] as documented in "
+                            "the UsdPreviewSurface and UsdUVTexture docs.",
+                            usdUvTextureShader.GetPath().GetText(),
+                            nonCompliantVector[0], nonCompliantVector[1],
+                            nonCompliantVector[2], nonCompliantVector[3]);
+    ValidateError(errors,
+        expectedErrorIdentifier,
+        usdUvTextureShader.GetPath(),
+        expectedErrorMsg);
```
Scope it and make expected* const.

tallytalwar replied:
Same below.

tallytalwar reviewed (commented):
Looks great, @beersandrew and very thorough test.

Just a few adjustments required for coding conventions, code cleanup etc.

beersandrew committed:
fix: addressing PR comments

- fixed many code styling bugs, including missing const's
- applied calls to ValidateError in all usdShade tests
- adding usdShadeShader schema to NormalMapTextureValidator
- fixing a CMake typo to include usdValidation

beersandrew commented on pxr/CMakeLists.txt:
```diff
@@ -8,7 +8,7 @@ if (EXISTS "${PROJECT_SOURCE_DIR}/pxr/exec")
 endif()
 
-if ($(PXR_BUILD_USD_VALIDATION))
+if (${PXR_BUILD_USD_VALIDATION})
```
@tallytalwar I believe this is a typo in `dev`

tallytalwar replied:
Yup @beersandrew , it was fixed with last dev push: https://github.com/PixarAnimationStudios/OpenUSD/commit/1f0b71a66ab76868d59739074d0a473b33ee0f3f

Can you update your dev branch?

beersandrew committed:
fix: styling fixes

- address tabs in error messages
- address long lines

beersandrew reviewed (commented):
@tallytalwar I believe all notes have been addressed

tallytalwar reviewed (approved):
Thanks all looks, good will sync this one in now.

beersandrew committed:
fix: update comment

tallytalwar commented on pxr/usdValidation/usdShadeValidators/testenv/testUsdShadeValidators.cpp:
```diff
@@ -536,22 +514,246 @@ TestUsdShadeEncapsulationRulesValidator()
+    usdPreviewSurfaceShaderPrim);
+        TF_AXIOM(errors.size() == 1u);
+
+        const TfToken expectedErrorIdentifier(
+        "usdShadeValidators:NormalMapTextureValidator.NonCompliantBiasAndScale");
+        const std::string expectedErrorMsg =
+            TfStringPrintf("UsdUVTexture prim <%s> reads 8 bit Normal Map "
+                           "@./normalMap.jpg@, which requires that "
+                           "inputs:scale be set to (2, 2, 2, 1) and "
+                           "inputs:bias be set to (-1, -1, -1, 0) for proper "
+                           "interpretation as per the UsdPreviewSurface and "
+                           "UsdUVTexture docs.",
+                           usdUvTextureShader.GetPath().GetText());
+        ValidatePrimError(errors[0],
+            expectedErrorIdentifier,
+            usdUvTextureShader.GetPath(),
+            expectedErrorMsg);
+    }
```
Looks like this one is failing:

Output from the ValidatePrimError seems to say that error identifier is file error and not NonCompliantBiasAndScale:

```
Error: usdShadeValidators:NormalMapTextureValidator.InvalidFile
Expected Error: usdShadeValidators:NormalMapTextureValidator.NonCompliantBiasAndScale
```

Can you make sure the test is passing on your end?

beersandrew replied:
This test is passing for me locally, unfortunately unable to reproduce what you're seeing.

The check for `InvalidFile` does happen before the `NonCompliantBiasAndScale` check so the if check [here](https://github.com/PixarAnimationStudios/OpenUSD/pull/3443/files#diff-4b963ec36ce7c8bdba6f4a2ed17d08b7f5422493898d17186304ec7eb9058e81R646) must be true for that test but I'm not sure why I'm not seeing the same

tallytalwar replied:
Ahh I see, weirdly the normalMap.jpg didn't get synced for me.

jessey-git commented:
This validator will end up flagging errors in https://github.com/usd-wg/assets/tree/main/test_assets/NormalsTextureBiasAndScale

That example uses scale-bias to adjust from DirectX to OpenGL normal map conventions for example. The validation error is noted there though, in the section for "r_normal_map_reversed_y.png". Would it be worthwhile to also allow the DirectX convention scale-bias to pass validation?

jesschimein commented:
/AzurePipelines run

tallytalwar commented:
Thanks @jessey-git for bringing this up. To begin with we can do the same treatment for directx related scale/bias values for normal maps as is being done for some scenarios here: https://github.com/PixarAnimationStudios/OpenUSD/commit/b718721744f72c310f9b08aeac101c9fecd9d89a

So instead of erroring, usdchecker will report these as warnings. Feel free to file a PR for this. Thanks
